src/java.desktop/share/classes/javax/swing/border/EtchedBorder.java

```diff
@@ -131,22 +151,37 @@ public EtchedBorder(int etchType, Color highlight, Color shadow)    {
      * @param height the height of the painted border
      */
     public void paintBorder(Component c, Graphics g, int x, int y, int width, int height) {
-        int w = width;
-        int h = height;
-
-        g.translate(x, y);
+        // We remove any initial transforms to prevent rounding errors
+        // when drawing in non-integer scales
+        AffineTransform at = new AffineTransform();
+        Stroke oldStk = new BasicStroke();
+        int stkWidth = 1;
+        if (g instanceof Graphics2D) {
+            at = ((Graphics2D) g).getTransform();
+            ((Graphics2D) g).setTransform(new AffineTransform());
+            oldStk = ((Graphics2D) g).getStroke();
+            stkWidth = (int) Math.floor(Math.min(at.getScaleX(),at.getScaleY()));
+            ((Graphics2D) g).setStroke(new BasicStroke((float) stkWidth));
+        }
 
-        g.setColor(etchType == LOWERED? getShadowColor(c) : getHighlightColor(c));
-        g.drawRect(0, 0, w-2, h-2);
+        int w = (int) Math.floor(at.getScaleX()*width-1);
+        int h = (int) Math.floor(at.getScaleY()*height-1);
 
-        g.setColor(etchType == LOWERED? getHighlightColor(c) : getShadowColor(c));
-        g.drawLine(1, h-3, 1, 1);
-        g.drawLine(1, 1, w-3, 1);
+        g.translate((int) Math.ceil(at.getScaleX()*x+at.getTranslateX()),
+                (int) Math.ceil(at.getScaleY()*y+at.getTranslateY()));
 
-        g.drawLine(0, h-1, w-1, h-1);
-        g.drawLine(w-1, h-1, w-1, 0);
+        paintBorderShadow(g, (etchType == LOWERED) ? getHighlightColor(c)
+                                                    : getShadowColor(c), w, h, stkWidth);
+        paintBorderRect(g, (etchType == LOWERED) ? getShadowColor(c)
+                                                    : getHighlightColor(c), w, h, stkWidth);
 
-        g.translate(-x, -y);
+        g.translate(-((int) (at.getScaleX()*x+at.getTranslateX())),
+                -((int) (at.getScaleY()*y+at.getTranslateY())));
```

Comment:
You used different values to move the translate at lines 170-171: `Math.ceil` is missing here. It's better to save the values above to make sure the translate is restored correctly.